packages/components/src/ui/context/context-system-provider.js

```diff
@@ -10,52 +10,68 @@ import {
 	createContext,
 	useContext,
 	useRef,
-	useState,
+	useEffect,
+	useMemo,
 	memo,
 } from '@wordpress/element';
-
-/**
- * Internal dependencies
- */
-import { useIsomorphicLayoutEffect } from '../utils';
+import warn from '@wordpress/warning';
 
 export const ComponentsContext = createContext(
 	/** @type {Record<string, any>} */ ( {} )
 );
 export const useComponentsContext = () => useContext( ComponentsContext );
 
+/**
+ * Runs an effect only on update (i.e., ignores the first render)
+ *
+ * @param {import('react').EffectCallback} effect
+ * @param {import('react').DependencyList} deps
+ */
+function useUpdateEffect( effect, deps ) {
+	const mounted = useRef( false );
+	useEffect( () => {
+		if ( mounted.current ) {
+			return effect();
+		}
+		mounted.current = true;
+		return undefined;
+	}, deps );
+}
+
 /**
  * Consolidates incoming ContextSystem values with a (potential) parent ContextSystem value.
  *
+ * Note: This function will warn if it detects an un-memoized `value`
+ *
  * @param {Object}              props
  * @param {Record<string, any>} props.value
  * @return {Record<string, any>} The consolidated value.
  */
 function useContextSystemBridge( { value } ) {
 	const parentContext = useComponentsContext();
-	const parentContextRef = useRef( parentContext );
-	const valueRef = useRef( merge( parentContext, value ) );
-
-	const [ config, setConfig ] = useState( valueRef.current );
 
-	useIsomorphicLayoutEffect( () => {
-		let hasChange = false;
+	const valueRef = useRef( value );
 
-		if ( ! isEqual( value, valueRef.current ) ) {
-			valueRef.current = value;
-			hasChange = true;
+	useUpdateEffect( () => {
+		if (
+			// objects are equivalent
+			isEqual( valueRef.current, value ) &&
+			// but not the same reference
+			valueRef.current !== value
+		) {
+			warn( `Please memoize your context: ${ JSON.stringify( value ) }` );
 		}
+	}, [ value ] );
 
-		if ( ! isEqual( parentContext, parentContextRef.current ) ) {
-			valueRef.current = merge( parentContext, valueRef.current );
-			parentContextRef.current = parentContext;
-			hasChange = true;
-		}
-
-		if ( hasChange ) {
-			setConfig( ( prev ) => ( { ...prev, ...valueRef.current } ) );
-		}
-	}, [ value, parentContext ] );
+	// parent context will always be memoized or the default value (which will not change)
```

Comment:
I'm not sure I follow completely the meaning of the first 2 sentences, would you mind articulating a bit better?

Reply:
I tried to re-write this/add more details, but I'm not sure it actually is helpful. Let me know what you think.

Reply:
Much better!

It's important that this function's implementation is well commented, since it's very dense and empowers the whole context system. Thank you!